pkg/projects/resource_project_role.go

```diff
@@ -0,0 +1,260 @@
+	"READ_POOLS_PIPELINE",
+	"MANAGE_INTEGRATIONS_PIPELINE",
+	"MANAGE_SOURCES_PIPELINE",
+	"MANAGE_POOLS_PIPELINE",
+	"TRIGGER_SECURITY",
+	"ISSUES_SECURITY",
+	"LICENCES_SECURITY",
+	"REPORTS_SECURITY",
+	"WATCHES_SECURITY",
+	"POLICIES_SECURITY",
+	"RULES_SECURITY",
+	"MANAGE_MEMBERS",
+	"MANAGE_RESOURCES",
+}
+
+type Role struct {
+	Name         string   `json:"name"`
+	Description  string   `json:"description"`
+	Type         string   `json:"type"`
+	Environments []string `json:"environments"`
+	Actions      []string `json:"actions"`
+}
+
+func (r Role) Id() string {
+	return r.Name
+}
+
+func (a Role) Equals(b Identifiable) bool {
+	return a.Id() == b.Id()
+}
+
+func rolesToEquatables(roles []Role) []Equatable {
+	var equatables []Equatable
+
+	for _, role := range roles {
+		equatables = append(equatables, role)
+	}
+
+	return equatables
+}
+
+func equatablesToRoles(equatables []Equatable) []Role {
+	var roles []Role
+
+	for _, equatable := range equatables {
+		roles = append(roles, equatable.(Role))
+	}
+
+	return roles
+}
+
+var unpackRoles = func(data *schema.ResourceData) []Role {
+	d := &ResourceData{data}
+
+	var roles []Role
+
+	if v, ok := d.GetOkExists("role"); ok {
+		projectRoles := v.(*schema.Set).List()
+		if len(projectRoles) == 0 {
+			return roles
+		}
+
+		for _, projectRole := range projectRoles {
+			id := projectRole.(map[string]interface{})
+
+			role := Role{
+				Name:         id["name"].(string),
+				Description:  id["description"].(string),
+				Type:         id["type"].(string),
+				Environments: castToStringArr(id["environments"].(*schema.Set).List()),
+				Actions:      castToStringArr(id["actions"].(*schema.Set).List()),
+			}
+			roles = append(roles, role)
+		}
+	}
+
+	return roles
+}
+
+var packRoles = func(d *schema.ResourceData, roles []Role) []error {
+	log.Printf("[DEBUG] packRoles")
+
+	setValue := mkLens(d)
+
+	var projectRoles []interface{}
+
+	for _, role := range roles {
+		log.Printf("[TRACE] %+v\n", role)
+		projectRole := map[string]interface{}{
+			"name":         role.Name,
+			"description":  role.Description,
+			"type":         role.Type,
+			"environments": role.Environments,
+			"actions":      role.Actions,
+		}
+
+		projectRoles = append(projectRoles, projectRole)
+	}
+
+	log.Printf("[TRACE] %+v\n", projectRoles)
+
+	errors := setValue("role", projectRoles)
+
+	return errors
+}
+
+var readRoles = func(projectKey string, m interface{}) ([]Role, error) {
+	log.Println("[DEBUG] readRoles")
+
+	roles := []Role{}
+
+	_, err := m.(*resty.Client).R().SetResult(&roles).Get(fmt.Sprintf(projectRolesUrl, projectKey))
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[TRACE] roles: %+v\n", roles)
+
+	// REST API returns all project roles, including ones with PREDEFINED type which can't be altered.
+	// We are only interested in the "CUSTOM" types that we can manipulate.
+	customRoles := roles[:0]
+	for _, role := range roles {
+		if role.Type == customRoleType {
```

Comment:
You should seperate this into it's own function for completeness sake - it's a filter